### PR TITLE
Add Gradle lockfiles and tool to generate them

### PR DESCRIPTION
**Pull request by blasten**
Related issue: https://github.com/flutter/flutter/issues/82745

Checks in the .lockfiles and adds a tool to generate them across the Flutter repo.

To regenerate these files, run `find . -type d -name 'android' | dart dev/tools/bin/generate_gradle_lockfiles.dart` 
from the root of the Flutter repo.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/benchmarks/complex_layout/android/build.gradle [hunk 1]
```diff
@@ -2,14 +2,24 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+// This file is auto generated.
+// To update all the build.gradle files in the Flutter repo,
+// See dev/tools/bin/generate_gradle_lockfiles.dart.
```
Comment (blasten) [inline dev/benchmarks/complex_layout/android/build.gradle:7]: The root build.gradle file will be auto-generated after this PR, and every plugin with Gradle dependency will also have its own lockfile. See https://github.com/flutter/flutter/tree/46d887f30003b5843911cf665af942c9c37fb23b/dev/benchmarks/complex_layout/android

File: dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig [hunk 2]
```diff
@@ -1 +1,2 @@
+#include? "Pods/Target Support Files/Pods-Runner/Pods-Runner.release.xcconfig"
```
Comment (dnfield) [inline dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig:1]: Seems unrelated to this PR.

File: dev/benchmarks/complex_layout/ios/Flutter/Debug.xcconfig [hunk 3]
```diff
@@ -1 +1,2 @@
+#include? "Pods/Target Support Files/Pods-Runner/Pods-Runner.debug.xcconfig"
```
Comment (dnfield) [inline dev/benchmarks/complex_layout/ios/Flutter/Debug.xcconfig:1]: Unrelated change.

File: dev/manual_tests/windows/flutter/generated_plugin_registrant.cc [hunk 4]
```diff
@@ -2,8 +2,6 @@
 //  Generated file. Do not edit.
 //
 
-// clang-format off
-
```
Comment (dnfield) [inline dev/manual_tests/windows/flutter/generated_plugin_registrant.cc:6]: unrelated change.

File: dev/manual_tests/windows/flutter/generated_plugin_registrant.h [hunk 5]
```diff
@@ -2,8 +2,6 @@
 //  Generated file. Do not edit.
 //
 
-// clang-format off
-
```
Comment (dnfield) [inline dev/manual_tests/windows/flutter/generated_plugin_registrant.h:6]: unrelated change.

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+// To regenerate these files, run `find . -type d -name 'android' | dart dev/tools/bin/generate_gradle_lockfiles.dart`
+
+import 'dart:io';
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:path/path.dart' as path;
+
+void main(List<String> arguments) {
+  print(
+      "Usage: find . -type d -name 'android' | dart dev/tools/bin/generate_gradle_lockfiles.dart\n\n");
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:16]: nit: no newline.

Comment (dnfield) [reply dev/tools/bin/generate_gradle_lockfiles.dart:16]: also nit: this makes it harder for developers on windows to run this script. Consider changing this to a file path argument where we just read the file.

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    if (!androidDirectory.childDirectory('app').existsSync()) {
+      print('${rootBuildGradle.path} is not an app - skipping');
+      continue;
+    }
+
+    if (!androidDirectory.parent.childFile('pubspec.yaml').existsSync()) {
+      print(
+          '${rootBuildGradle.path} no pubspec.yaml in parent directory - skipping');
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:46]: no newline

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    print('Processing ${androidDirectory.path}');
+
+    rootBuildGradle.writeAsStringSync(rootGradleFileContent);
+    settingsGradle.writeAsStringSync(settingGradleFile);
+
+    // Fetch pub dependencies.
+    exec('flutter', <String>['pub', 'get'],
+        workingDirectory: androidDirectory.parent.absolute.path);
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:65]: style guide says this should either be one line or all new lines with a trailing comma

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+    // Verify that the Gradlew wrapper exists.
+    final File gradleWrapper = androidDirectory.childFile('gradlew');
+    // Generate Gradle wrapper if it doesn't exists.
+    // This logic is embededed within the Flutter tool.
+    // To generate the wrapper, build a flavor that doesn't exist.
+    if (!gradleWrapper.existsSync()) {
+      Process.runSync('flutter',
+          <String>['build', 'apk', '--debug', '--flavor=does-not-exist'],
+          workingDirectory: androidDirectory.parent.absolute.path);
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:75]: ditto

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    // Generate lock files.
+    exec(gradleWrapper.absolute.path, <String>[':generateLockfiles'],
+        workingDirectory: androidDirectory.absolute.path);
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:80]: ditto

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  return files;
+}
+
+void exec(
+  String cmd,
+  List<String> args, {
+  String? workingDirectory,
+}) {
+  final ProcessResult result =
+      Process.runSync(cmd, args, workingDirectory: workingDirectory);
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:104]: no newline

File: dev/tools/bin/generate_gradle_lockfiles.dart [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+  if (result.exitCode != 0 || '${result.stderr}'.isNotEmpty)
+    throw ProcessException(
+        cmd, args, '${result.stdout}${result.stderr}', result.exitCode);
```
Comment (dnfield) [inline dev/tools/bin/generate_gradle_lockfiles.dart:107]: ditto

**Review by dnfield — COMMENTED** [review]
Style nits. I think making the script work on Windows would be beneficial.

Please revert the unrelated files (*.xcconfig, *.h, *.cc).

Consider landing this in two steps: one where we just update all the gradle deps, and one where we actually add the lock files (or vice versa). Right now doing both may make it harder to bisect if anything goes wrong in devicelab.

File: dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig (hunk 2 above)
Comment (blasten) [reply dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig:1]: @jmagman would it be ok to add this file to .gitignore?

Comment (blasten) [reply dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig:1]: These files were created after `flutter pub get`, which is required prior to creating the lockfiles.

Comment (dnfield) [reply dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig:1]: Maybe we should just go through and update all of these then in a separate pr.  Or alternatively gitignore them and remove them

File: dev/benchmarks/complex_layout/ios/Flutter/Debug.xcconfig (hunk 3 above)
Comment (blasten) [reply dev/benchmarks/complex_layout/ios/Flutter/Debug.xcconfig]: Reverted. Done

File: dev/manual_tests/windows/flutter/generated_plugin_registrant.h (hunk 5 above)
Comment (blasten) [reply dev/manual_tests/windows/flutter/generated_plugin_registrant.h]: Done

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 6 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: what would be the issue?  I have added a section about how to enter the files manually to write to the stdin pipe.

FWIW,  the blaze command (internally) works this way for running multiple targets at once. I don't know what part wouldn't work on Windows since powershell also has support for the pipe operator.

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 7 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: fixed. This was caused by dartfmt

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 8 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: Done

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 10 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: Done

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 9 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: Done

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 11 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart:112]: Done

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 12 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart:115]: Done

File: dev/manual_tests/windows/flutter/generated_plugin_registrant.cc (hunk 4 above)
Comment (blasten) [reply dev/manual_tests/windows/flutter/generated_plugin_registrant.cc:6]: Done

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 6 above)
Comment (dnfield) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: Ahh ok, so I guess on Windows you could just use something like `gci -Recurse -Directory -Name "android" | dart ...`. That should be fine then.

**Review by dnfield — APPROVED** [review]
LGTM

File: dev/tools/bin/generate_gradle_lockfiles.dart (hunk 6 above)
Comment (blasten) [reply dev/tools/bin/generate_gradle_lockfiles.dart]: 🎉

**Comment by blasten** [comment]
@godofredoc I'd like to see if this change also makes Gradle builds fasters. Is there a SQL I could run to see the before/after of this change?

**Comment by blasten** [comment]
@godofredoc I'd like to see if this change also makes Gradle build fasters. Is there a SQL I could run to see the before/after of this change?

File: dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig (hunk 2 above)
Comment (jmagman) [reply dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig]: These xcconfigs should be checked into the project, not gitignored.  I'm not sure why it's not already for `dev/benchmarks/complex_layout`.  I can look into that in another PR.

File: dev/benchmarks/complex_layout/ios/Flutter/Debug.xcconfig [hunk 13]
```diff
@@ -1 +0,0 @@
-#include "Generated.xcconfig"
```
Comment (jmagman) [inline dev/benchmarks/complex_layout/ios/Flutter/Debug.xcconfig:1]: Don't delete this file.

File: dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig [hunk 14]
```diff
@@ -1 +0,0 @@
-#include "Generated.xcconfig"
```
Comment (jmagman) [inline dev/benchmarks/complex_layout/ios/Flutter/Release.xcconfig:1]: Or this one.

**Review by jmagman — CHANGES_REQUESTED** [review]

**Review by jmagman — APPROVED** [review]
LGTM

When will generate_gradle_lockfiles need to be run?

**Comment by blasten** [comment]
@jmagman it would need to be run if a Gradle dependencies is added or modified from these apps.  That includes modifying Flutter plugins.  I've added a custom Gradle error to https://github.com/flutter/flutter/pull/83067/files#diff-c0a1cddc5ca2d811173e4bd32e800ed637b1203c37a22907703b2eaa00899dbb to guide folks on how to resolve this issue on their own.

**Comment by blasten** [comment]
Here: https://github.com/flutter/flutter/blob/a7cd63eba3189cc89cea5f0cf1b795849a28321a/packages/flutter_tools/lib/src/android/gradle_errors.dart#L436-L459
